packages/@aws-cdk/core/lib/tree.ts

```diff
@@ -0,0 +1,47 @@
+/**
+ * Interface for inspecting a construct and adding metadata into the construct tree
+ */
+export interface ITreeInspector {
+  /**
+   * Adds key-value pairs of metadata as attributes
+   *
+   * @param key - key for metadata
+   * @param value - value of metadata
+   */
+  attribute(key: string, value: any): void;
+}
+
+/**
+ * Inspector that maintains an attribute bag
+ */
+export class TreeInspector implements ITreeInspector {
+
+  /**
+   * Represents the bag of attributes as key-value pairs.
+   */
+  public readonly attributes: Record<string, any> = {};
```

Comment:
What's `Record`?

Reply:
An initial thought was restricting the keys to ones we've defined by convention  - [Record](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt)

I think the more idiomatic thing if we want to accept *any* string as a key would be to just go with `{ [key:string]:any }`